### PR TITLE
Add automatic table header formatting

### PR DESCRIPTION
**Pull request by rahlvers**

### CODE REVIEW

**Comment by mraineri** [comment]
@akf please review this when you have a chance. If it looks good to you we'll merge this in.

**Comment by akf** [comment]
Unfortunately there have been enough changes in the master branch since Jack worked on this that there are merge conflicts. These are pretty easily resolved. (I just did so, but I've just realized I'm not sure how to get them into this pull request without completing the merge to master!) 

There are two other issues, no fault of Jack's: 

1) As of a few weeks ago, any text strings the doc generator emits should be wrapped for localization. (This means replacing 'Some text string' with _('Some text string') and, ideally, using placeholder syntax in strings rather than building them up through concatenation. I can make these changes; it's just a bit more of a manual review as most of the additions didn't trigger merge conflicts. I'll also regenerate the doc_generator.pot file from the resulting files. 

2) This broke some of the tests. This is no surprise, as a lot of the tests check for specific output that has been changed here on purpose. 

The above are things I was expecting, and are why I suggested we add an item to the next SOW for me to work through this.

Most importantly, there does seem to be a bug: 

>       table_id = self.get_next_table_id();
E       AttributeError: 'MarkdownGenerator' object has no attribute 'get_next_table_id'
doc_formatter/doc_formatter.py:456: AttributeError 

I'm getting this when I try to generate HTML output too. It's possible I messed up something in my merge, but I don't think so.

File: doc-generator/doc_formatter/doc_formatter.py [hunk 1]
```diff
@@ -433,9 +438,13 @@ def format_conditional_details(self, schema_ref, prop_name, conditional_reqs):
                     req += ', must be ' + comparison + ' ' + ', '.join(['"' + val + '"' for val in values])
             rows.append(self.formatter.make_row([req_desc, req, purpose]))
 
+        table_id = self.get_next_table_id();
```
Comment (akf) [inline doc-generator/doc_formatter/doc_formatter.py:441]: get_next_table_id isn't defined. Probably lost during a merge at some point?

**Review by akf — CHANGES_REQUESTED** [review]
We definitely need that get_next_table_id function. 

I added some other comments on the pull request -- there's work to be done getting the new strings wrapped for localization (small stuff) and getting the tests to pass, where the changes in output are intentional.

**Comment by rahlvers** [comment]
@akf - Amy, I've tagged Jack (@sbs-jms ) to take a look at some fixes.

**Comment by sbs-jms** [comment]
I just checked my local copy of the code, and while everything is checked in, it may not have been pushed when the merge was done.

At this point, get_next_table_id() is no longer needed, as the new code is not producing table numbers, but merely placeholders where numbers will be needed (i.e.., "Table TBL_NN++", rather than "Table 7"). The placeholders are then used by an awk script to do the actual substitution, once all of the spec content has been generated, and the sequencing is straightforward.

I'll take a look at the code for any other "leftovers", and will put together a summary of the changes/approach, since what has worked for us with SSM may differ from what Redfish needs more broadly.

**Comment by sbs-jms** [comment]
@akf I've removed the spurious get_next_table_id() calls. They dated from a prior attempt and were still lingering in routines/subsitution points that Swordfish didn't use. I also added a little more explanation of our approach to numbering. Let me know if there is more I can do.

**Comment by akf** [comment]
Thanks, @sbs-jms!

@mraineri -- please read through the TABLE_NUMBER_README.md @sbs-jms has added here, and confirm whether this is the behavior you want the main line of the doc generator to have going forward. (I would expect you might want it to be optional.) If it is, I'll give this another review.

**Comment by mraineri** [comment]
@jautor can you take a look at this? If the behavioral change is okay, then I'll work on merging this.

**Comment by mraineri** [comment]
@akf we're okay with the changes being proposed, but we need some help merging this in without breaking items you've been working on. Would you be able to manage this?

**Comment by akf** [comment]
I'm out of the office until the 21st (peeking at my email, though), but
yes. I'll take a look when I get back.

**Comment by akf** [comment]
I've merged this into a separate branch and added a draft pull request, #330 , for reference. In addition to resolving the merge conflicts, I've added the translation function wrapper around the newly introduced strings, and updated the affected tests. 

I see one issue I don't know how to resolve: Headings have been added for the "release history" tables, but they don't really make sense. This update adds a header for two columns, labeling them "Version" and "Release" -- regardless of whether there are one, two, or more versions to output.  So we get something like this: 

| Version | Release |
| --- | --- | --- | --- | --- | --- | --- |
| *v1.6* | *v1.5* | *v1.4* | *v1.3* | *v1.2* | *v1.1* | *v1.0* |
| 2018.3 | 2018.2 | 2017.3 | 2017.2 | 2017.1 | 2016.2 | 2016.1 |

It would make more sense to do this, but I suspect that doesn't pass muster with ISO: 

| --- | --- | --- | --- | --- | --- | --- | --- |
| Version | *v1.6* | *v1.5* | *v1.4* | *v1.3* | *v1.2* | *v1.1* | *v1.0* |
| Release | 2018.3 | 2018.2 | 2017.3 | 2017.2 | 2017.1 | 2016.2 | 2016.1 |

What we're presenting here isn't really tabular information, of course; this is just a compact way of displaying a list of versions. 

Ideas?

**Review by akf — APPROVED** [review]

**Comment by sbs-jms** [comment]
@akf : I think your rework of the version table is the right thing to do, and won't cause any ISO issues. Generally, ISO's concern is that:
- any and all content be uniquely identifiable. This is basis of their dislike of  "hanging paragraphs" with content in both clause x and clause x.y (If I complain about an error in clause x, am I referring to just clause x or to clause x and all its sub-clauses?), and the need for table captions and references, and 
- a clear distinction be made between normative content (the "shalls") and informative content

They are not, as a rule, concerned with formatting issues, and then only in the interest of further clarity and usefulness. Your proposed rework of the version table makes the content clearer and more useful without violating any higher-order ISO-isms. I'm all for it.

**Comment by akf** [comment]
Thanks, @sbs-jms! I'm making Version/Release "labels" in the first column of the release history tables.

**Comment by akf** [comment]
Okay, this is all resolved and merged. Thanks, everyone!

**Pull request merged**